### ISSUE TITLE
Encoding / Concatinating the Global Planner GPS Goal locations

### ISSUE THREAD

**Issue by Alobaali12** [autonomousvision/carla_garage#64]
Hi, 

I hope my message finds you well. 

I would like to ask why a GRU is specifically used to help with waypoint predictions in transfuer ? 

<img width="452" alt="Image" src="https://github.com/user-attachments/assets/b6b1a52d-d49d-4f51-a2ed-a2804e22d917" />

I would like some advice with regards to the below:

I'm training a classical convolutional neural network that takes in input images from PDM-Lite , goes through the CNN network, Flattened and then enters a MLP, I would like to add to my network information about the goal locations, current speed, etc.. So I concatenate (Goal location, current speed) to the hidden layers in my MLP that processed the feature vector. 

Is that correct, can I just add the new information to the MLP layers this way, is there any drawbacks, will it work, and do you have any recommendation ? 

One problem is how do I ensure the size of my input data (goals and current speed) does not negatively affect the network at that later stage, for example they will be different to the size of my processed feature map.

**Comment by Kait0** [comment]
I think this GRU for waypoints was originally developed in motion forecasting literature.
I have not extensively ablated their impact.

Adding values to an MLP is fine. We sometimes encode them first with their own MLP before concatenation.
You can and probably should normalize your extra inputs.
The high-effort way is to compute the mean and standard deviation over your dataset and normalize using that.
The low effort way is to add a batch norm as a first layer, which does the same.

As we show in the hidden biases paper you might want to use a Transformer instead of MLP if you care about performance. You can add the extra information as another token than like we do in our method.

**Comment by Alobaali12** [comment]
thank you
